openassessment/xblock/static/js/spec/lms/oa_response.js

```diff
@@ -583,6 +581,76 @@ describe("OpenAssessment.ResponseView", function() {
+
+        expect(view.submitEnabled()).toBe(false);
+
+        // Upload files
+        var files = [{type: 'image/jpeg', size: 1024, name: 'picture1.jpg', data: ''},
+                     {type: 'image/jpeg', size: 1024, name: 'picture2.jpg', data: ''}];
+        view.prepareUpload(files, 'image', ['i1', 'i2']);
+        view.uploadFiles()
+        view.checkSubmissionAbility(true);
+        expect(view.submitEnabled()).toBe(true);
+        expect(view.files.length).toEqual(2);
+        view.prepareUpload(files, 'image', ['i1', 'i2']);
+        view.uploadFiles()
+        view.checkSubmissionAbility(true);
+        expect(view.files.length).toEqual(2);
+    });
+
+    it("tests that file upload number can't exceed maximum", function() {
+        spyOn(view.baseView, 'toggleActionError').and.callThrough();
+        spyOn(view, 'getSavedFileCount').and.returnValue(20);
+        view.textResponse = 'optional';
+        view.fileUploadResponse = 'required';
+        expect(view.submitEnabled()).toBe(false);
+        var files = [];
+        for(var i=0; i<20;i++) {
+            files.push({type: 'image/jpeg', size: 1024, name: 'picture1.jpg', data: ''});
+        }
+
+        view.prepareUpload(files, 'image', ['i1', 'i2','i1', 'i2','i1', 'i2','i1', 'i2','i1', 'i2','i1', 'i2',
+        'i1', 'i2','i1', 'i2','i1', 'i2','i1', 'i2',]);
+        view.checkSubmissionAbility(true);
+        expect(view.submitEnabled()).toBe(true);
+        expect(view.getSavedFileCount()).toEqual(20);
+        var files2 = [];
+        for(i=0; i<2;i++) {
+            files2.push({type: 'image/jpeg', size: 1024, name: 'picture1.jpg', data: ''});
+        }
+        view.prepareUpload(files2, 'image', ['i1', 'i2']);
+        expect(view.baseView.toggleActionError).toHaveBeenCalledWith(
+            'upload', 'Only ' + view.data.MAXIMUM_FILE_UPLOAD_COUNT +' files can be saved.');
+    });
+
+    it("tests that file upload works after file delete", function() {
+        view.textResponse = 'optional';
+        view.fileUploadResponse = 'required';
+
+        expect(view.submitEnabled()).toBe(false);
+
+        // Upload files
+        var files = [{type: 'image/jpeg', size: 1024, name: 'picture1.jpg', data: ''},
+                     {type: 'image/jpeg', size: 1024, name: 'picture2.jpg', data: ''}];
+        view.prepareUpload(files, 'image', ['i1', 'i2']);
+        view.uploadFiles()
+        view.checkSubmissionAbility(true);
+        expect(view.submitEnabled()).toBe(true);
+        // Delete the first file
+        view.removeUploadedFile(0);
+        view.checkSubmissionAbility(true);
+        expect(view.submitEnabled()).toBe(true);
+        files = [{type: 'image/jpeg', size: 1024, name: 'picture3.jpg', data: ''}];
+        view.prepareUpload(files, 'image', ['i3']);
+        view.uploadFiles();
+        view.checkSubmissionAbility(true);
+        expect(view.fileNames.length).toEqual(1);
+        expect(view.filesDescriptions.length).toEqual(1);
```

Comment:
This doesn't assert anything about the state of the previously uploaded files.  The issue we ran into during manual testing wasn't that uploads didn't work, it's that the file description of the newly uploaded file was something unexpected.

Reply:
The root cause of the issue was that the fileNames collection wasn't being cleared and we were appending elements to it. The fix in the code, ensured that the fileNames collection is cleared after each upload and its length should be the same as fileDescriptions collection which is the same as the number of files selected for upload. Hence my approach to testing it this way. 
Another reason for this approach: we "spy" on the saveFilesDescriptions method - which is where the fileNames collection is supposed to be . populated. I feel this testing is sufficient and full fills the purpose.

Reply:
OK, that's a fair enough argument.  Could you leave a comment to the effect of
```ensure that the fileNames collection is cleared after each upload, and that its length is the same as the fileDescriptions collection (and the number of files selected for upload). ```
before these two assertions?

Reply:
UT updated

Reply:
How about
```js
// Ensure that view.fileNames and view.fileDescriptions only contain data about the newest
// set of uploaded files.
expect(view.fileNames).toEqual(['picture3.jpg']);
expect(view.filesDescriptions).toEqual(['i3']);
```